File: sample.lua
print("Hello World!")
require "math"
enemy = {
    position = {
        x = 200.0,
        y = 300.0,
        z = 400.0,
    },
}

function enemy.wander (player) 
    print("lua recieved", player["x"], player["y"], player["z"]);
    x = math.random(enemy["position"]["x"] - 100, enemy["position"]["x"] + 100)
    y = math.random(enemy["position"]["y"] - 100, enemy["position"]["y"] + 100)
    z = math.random(enemy["position"]["z"] - 100, enemy["position"]["z"] + 100)

    enemy["position"]["x"] = x
    enemy["position"]["y"] = y
    enemy["position"]["z"] = z

    return enemy["position"]
end

File: sample.cpp
#include "luabind.hpp"
class Sample {
    public:
        Sample(std::string script);
        ~Sample();

        void runluafunction(std::string function, glm::vec3 p);
        void getluaglobal(std::string global);
        LuaBind * luaBind;
};

Sample::Sample(std::string script) {
    luaBind = new LuaBind(script);
}

Sample::~Sample() {
    luaBind->clean();
    delete luaBind;
}

void Sample::getluaglobal(std::string global) {
    glm::vec3 p = luaBind->get<glm::vec3>(global);
    std::cout << "C++ recieved " << p.x << ", " << p.y << ", " << p.z << std::endl;
}

void Sample::runluafunction(std::string function, glm::vec3 p) {
    int numargs = 1;
    int numresults = 1;
    glm::vec3 args[numargs] = {p};
    luaBind->execute(function, numargs, args, numresults);
    glm::vec3 r = luaBind->getTop<glm::vec3>();
    std::cout << "C++ recieved " << r.x << ", " << r.y << ", " << r.z << std::endl;

}


int main() {
    Sample *s = new Sample("sample.lua");
    s->getluaglobal("enemy.position");
    s->runluafunction("enemy.wander", glm::vec3(1));

    delete s;
    return 0;
}
